src/Elmish/Hooks/UseRef.purs

```diff
@@ -0,0 +1,44 @@
+module Elmish.Hooks.UseRef
+  ( Ref
+  , useRef
+  )
+  where
+
+import Prelude
+
+import Data.Maybe (Maybe(..))
+import Data.Tuple.Nested (type (/\), (/\))
+import Elmish (EffectFn1, (<?|))
+import Elmish.Hooks (UseState, useState)
+import Elmish.Hooks.Type (Hook)
+import Elmish.Hooks.Type as Hooks
+import Web.HTML (HTMLElement)
+
+type UseRef = UseState (Maybe HTMLElement)
+
+type Ref = Maybe HTMLElement /\ (EffectFn1 HTMLElement Unit)
+
+-- | The `useRef` hook returns a `Hook` encapsulating an element paired with a
+-- | setter for that element. This setter can be passed to a `ref` prop to get a
+-- | reference to an element, but it has to be `unsafeCoerce`d because
+-- | `elmish-html` expects `ref` to be a string.
+-- |
+-- | ```purs
+-- | view :: ReactElement
+-- | view = withHooks do
+-- |   inputEl /\ setInputEl <- useRef Nothing
+-- |   let onButtonClick = traverse_ focus inputEl
+-- |   pure $
+-- |     H.fragment
+-- |     [ H.input_ "form-control" { ref: unsafeCoerce setInputEl, defaultValue: "" }
```

Comment:
How about giving the `ref` prop on `elmish-html` the right type so you don't have to use `unsafeCoerce`?

Reply:
Yeah, but what do we want to say is the right type? `String` is allowed, but I think deprecated, so probably the worst choice. But then there’s the `{ current :: a }` thing which is used by `useRef` and `createRef`. Should we say that this is the officially supported way?

Reply:
Oh, maybe it could be an opaque type and `elmish-html` could export a `functionRef :: EffectFn1 HTMLElement Unit -> Ref`?

Reply:
Well, first, I do think the function should be the officially supported way, because it's, well, _functional_. The container object requires an effect for initialization, so it will always be awkward to use.

But yes, opaque type is also a fine idea.

Reply:
Okay, to keep it simple, then, I’ll start with `EffectFn1 HTMLElement Unit` being the supported type 👍🏻

Reply:
I looked at the docs on this, apparently `createRef` & `useRef` are newer. The callback ref approach has the benefit of supporting more versions of React. 

<img width="902" alt="Screen Shot 2022-04-13 at 4 25 26 PM" src="https://user-images.githubusercontent.com/102529518/163264713-a127859f-cd6a-4210-927b-5b1f0b809fce.png">
From: https://reactjs.org/docs/refs-and-the-dom.html#dont-overuse-refs

Reply:
String refs are definitely deprecated, like Marc said.
https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs

Reply:
> Oh, maybe it could be an opaque type and `elmish-html` could export a `functionRef :: EffectFn1 HTMLElement Unit -> Ref`? Then we could use it here and return a `Hook UseRef (Maybe HTMLElement /\ Elmish.HTML.Ref)`.

It sounds like this approach would preserve the flexibility that React refs have in the underlying framework.

Like instead of picking one opinionated way of creating refs, I vote having methods like this one to create opaque refs. 
It's less churn the next time React deprecates an api or adds a new method to create refs. There's about 4 ways to make one right now.

Reply:
Alright, yeah, that makes sense. For now I’ll make an opaque type and one way to create that type (`Elmish.HTML.callbackRef`).

Reply:
In that case, I think the opaque type should go into `elmish`, not `elmish-html`. And since we already have a type named `Ref` (with completely different semantics), I think we need to rename it. I propose `Opaque`. I think it's ok if it's not immediately intuitive: since the concept behind it is also not readily obvious, it would take some studying to understand it anyway.

And the new type should go, I think, under `Elmish.React`. But still reexported from the root.

Reply:
Now I’m wondering if it can even work. Think about the way you deal with each type: For `callbackRef`, you have a separate value in the state that gets set. For `createRef`, the value is the same as the one you pass to the `ref` prop. So that means you need a `readRef :: Ref -> Maybe HTMLElement`, right? But that doesn’t make sense for `Ref`s created via `callbackRef`. I guess it could still return `Nothing`, but it just seems weird that a function could get passed to `readRef`. I’m leaning towards just doing `EffectFn1 HTMLElement Unit` for now and punting on deciding what the type should be as React evolves.

Reply:
For `createRef` there would just be an intermediate type.

```
createRef :: Effect CellRef
readRef :: CellRef -> Maybe Element

cellRef :: CellRef -> Ref
callbackRef :: Dispatch Element -> Ref
```

But yes, this kind of awkwardness is exactly why I didn't want to do the opaque type in the first place.

Reply:
Oh, and by the way: the type will have to be generic, i.e. `Ref a`, because it's not only `HTMLElement`s that can be accessed through it. Just yesterday, Charles used it to call methods on the slider component.

Reply:
Ohh that is a good idea. But yeah, I think I’d rather defer for now and make a slight improvement rather than figure out what the best way to represent all this is.

Reply:
And the genericity requirement applies to the naked function approach too

Reply:
Oh right, well then maybe I should just make it `Foreign` like our event handler types?

Reply:
Yeah, I guess that makes sense for now

Reply:
But then we’re right back to `unsafeCoerce`-ing 🤔 What’s your thinking on making it generic? That’s very similar to `unsafeCoerce` in practice, because the caller can make it any type they want.

Reply:
Well, no. If it's generic, then it's the props of the component that get to choose the type. So, for example:

```
type OptProps_input =
  { ...
    ...
    ref :: EffectFn1 HTMLInputElement Unit
    ...
  }

type OptProps_div =
  { ...
    ...
    ref :: EffectFn1 HTMLDivElement Unit
    ...
  }
```

And any custom component would just make their own opaque type for this:

```
foreign import data MyComponent

type SliderProps = 
  { ...
    ref :: EffectFn1 MyComponent Unit
    ...
  }

foreign import myComponentMethod :: MyComponent -> Effect Unit
```

Reply:
Ah, I see what you’re saying. So `useRef` is generic. That makes sense — I’ll do that 👍🏻

Reply:
Although it makes it very difficult to work with the code-gen in `elmish-html`. I’ll think about it some more.

Reply:
If everything is named sanely, we can just to `Web.HTML.HTML{Capitalizedtag}Element`